piptools/writer.py

```diff
@@ -28,7 +28,7 @@ def __init__(self, src_files, dst_file, dry_run, emit_header, emit_index,
     def _sort_key(self, ireq):
         return (not ireq.editable, str(ireq.req).lower())
 
-    def write_header(self):
+    def write_header(self, allow_unsafe):
```

Comment:
Maybe it would be better to injected `allow_unsafe` to the `OutputWriter` constructor instead?

Reply:
Certainly can do that. 

Right now `allow_unsafe` gets here from being passed to each call of `OutputWriter.write`. Is there any big benefit from the current model? I could see re-using a single instance across `write` calls where this value varies, if it's slow to construct the instance. It looks like both the constructor and `write` take a lot of arguments, and I'm not sure how they are meant to be split; however, I only see one caller of `write`, which is right after instantiation, so maybe there isn't an intentional divide...

Ah, the other caller is in the tests, so I'll need either a new "unsafe writer" fixture, or I can update the tests to expect the fixture gives them a class (or partial), instead of an instance. Or maybe there's a third, better way? Preferences there?

Reply:
My point is it looks a bit hacky, because the `allow_unsafe` injected now across two layers `_iter_lines` and `write_header` and it could go deeper. A quick research showed **it was already in the constructor earlier** (see #377 PR) and later was deleted by @jdufresne (see 68533537 commit). Current `allow_unsafe` was appeared in `_iter_lines` (see #517 PR).

Of course it'll affect on tests, but we could create writer instance and write a test for each behavior, allow_unsafe or not allow_unsafe, separately which looks better to me.

Reply:
Now it looks pretty good to me. Thanks!